fixed-hash/src/hash.rs

```diff
@@ -6,735 +6,646 @@
+					return $crate::core::cmp::Ordering::Greater;
 				}
-				ret
+				$crate::core::cmp::Ordering::Equal
 			}
 		}
 
-		#[deprecated(
-			since = "0.2.5",
-			note = "misses proper error handling; will be replaced by `new_from_slice` in version 0.3"
-		)]
-		impl<'a> From<&'a [u8]> for $from {
-			fn from(s: &'a [u8]) -> $from {
-				$from::from_slice(s)
+		#[cfg(any(not(feature = "libc"), target_os = "unknown"))]
+		impl $crate::core::cmp::PartialEq for $name {
+			#[inline]
+			fn eq(&self, other: &Self) -> bool {
+				self.as_bytes() == other.as_bytes()
 			}
 		}
 
-		impl_std_for_hash!($from, $size);
-		impl_heapsize_for_hash!($from);
-		impl_libc_for_hash!($from, $size);
-		impl_quickcheck_arbitrary_for_hash!($from, $size);
-	}
-}
-
-/// Implements conversion to and from hash types of different sizes. Uses the
-/// last bytes, e.g. `From<H256> for H160` uses bytes 12..32
-/// CAUTION: make sure to call with correct sizes and the bigger type first or
-/// bad things will happen!
-#[macro_export]
-macro_rules! impl_hash_conversions {
-	($a: ident, $a_size: expr, $b: ident, $b_size: expr) => {
-		impl From<$b> for $a {
-			fn from(value: $b) -> $a {
-				debug_assert!($a_size > $b_size && $a_size % 2 == 0 && $b_size %2 == 0);
-				let mut ret = $a::new();
-				ret.0[($a_size - $b_size)..$a_size].copy_from_slice(&value);
-				ret
+		#[cfg(any(not(feature = "libc"), target_os = "unknown"))]
+		impl $crate::core::cmp::Ord for $name {
+			#[inline]
+			fn cmp(&self, other: &Self) -> $crate::core::cmp::Ordering {
+				self.as_bytes().cmp(other.as_bytes())
 			}
 		}
 
-		impl From<$a> for $b {
-			fn from(value: $a) -> $b {
-				debug_assert!($a_size > $b_size && $a_size % 2 == 0 && $b_size %2 == 0);
-				let mut ret = $b::new();
-				ret.0.copy_from_slice(&value[($a_size - $b_size)..$a_size]);
+		#[cfg(feature = "rand-support")]
+		impl $crate::rand::distributions::Distribution<$name>
+			for $crate::rand::distributions::Standard
+		{
+			fn sample<R: $crate::rand::Rng + ?Sized>(&self, rng: &mut R) -> $name {
+				let mut ret = $name::zero();
+				for byte in ret.as_bytes_mut().iter_mut() {
+					*byte = rng.gen();
+				}
 				ret
 			}
 		}
 
-		impl<'a> From<&'a $b> for $a {
-			fn from(value: &'a $b) -> $a {
-				let mut ret = $a::new();
-				ret.0[($a_size - $b_size)..$a_size].copy_from_slice(value);
-				ret
+		/// Utilities using the `rand` crate.
+		#[cfg(feature = "rand-support")]
+		impl $name {
+			/// Assign `self` to a cryptographically random value using the
+			/// given random number generator.
+			pub fn randomize_using<R>(&mut self, rng: &mut R)
+			where
+				R: $crate::rand::Rng + ?Sized
+			{
+				use $crate::rand::distributions::Distribution;
+				*self = $crate::rand::distributions::Standard.sample(rng);
 			}
-		}
-	}
-}
 
-/// Implements conversion to and from a hash type and the equally sized unsigned int.
-/// CAUTION: Bad things will happen if the two types are not of the same size!
-#[cfg(feature="uint_conversions")]
```

Comment:
What is the plan for these? Move them to `uint`? Own crate? Just remove them? They are currently in use in `ethereum-types` so we need some kind of upgrade path for consumers.

Reply:
I had the idea to move them to `pwasm-std` for the `pwasm-*` crates and also provide `uint` types by `pwasm-std` so it fits again for it. And for `ethereum-types` I would do the same I guess. Maybe just another crate to unify both? It actually is not feasible to make `uint` depending on `fixed-hash` since it is built up from `u64` words instead of bytes and would require another major overhaul. However, `uint` is far more complex than `fixed-hash`.

So to summarize: The best way is to have this functionality in the crate that actually needs this interoperability **or** in another small utility crate featuring both types.

Reply:
Hmm, a tricky one! My personal opinion is that it's probably less disruptive/more pragmatic to just keep these conversions in `fixed-hash` for the time being and refactor later – maybe we can make the consuming code use the conversions less or not at all and we can just remove this? If you feel strongly this is not a good path ahead, I can be swayed though (though we can't merge before we have a solution to this as it'd make upgrading impossible).